#!/usr/bin/env python

# Software License Agreement (BSD License)
#
# Copyright (c) 2012, Robotiq, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of Robotiq, Inc. nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
# Copyright (c) 2012, Robotiq, Inc.
# Revision $Id$

"""@package docstring
Command-line interface for receiving and interpreting the status of a vacuum gripper.

This serves as an example for receiving messages from the 'RobotiqVacuumGrippersRobotInput' topic using the 'RobotiqVacuumGrippers_robot_input' msg type and interpreting the corresponding status of the vacuum gripper.
"""

import roslib; roslib.load_manifest('robotiq_vacuum_grippers_control')
import rospy
from std_msgs.msg import String
from robotiq_vacuum_grippers_control.msg import _RobotiqVacuumGrippers_robot_input  as inputMsg

def printStatus(status):
    """Print the status string generated by the statusInterpreter function."""

    print statusInterpreter(status)

def RobotiqVacuumGrippersStatusListener():
    """Initialize the node and subscribe to the RobotiqVacuumGrippersRobotInput topic."""

    rospy.init_node('RobotiqVacuumGrippersStatusListener')
    rospy.Subscriber("RobotiqVacuumGrippersRobotInput", inputMsg.RobotiqVacuumGrippers_robot_input, printStatus)
    rospy.spin()

def statusInterpreter(status):
    """Generate a string according to the current value of the status variables."""

    output = '\n-----\nVacuum gripper status interpreter\n-----\n'

    #gACT
    output += 'gACT = ' + str(status.gACT) + ': '
    if(status.gACT == 0):
       output += 'Gripper reset\n'
    if(status.gACT == 1):
       output += 'Gripper activation\n'

    #gMOD
    output += 'gMOD = ' + str(status.gMOD) + ': '
    if(status.gMOD == 0):
        output += 'Gripper set to automatic mode\n'
    if(status.gMOD == 1):
        output += 'Gripper set to advanced mode\n'

    #gGTO
    output += 'gGTO = ' + str(status.gGTO) + ': '
    if(status.gGTO == 0):
        output += 'Standby (or performing activation/automatic release)\n'
    if(status.gGTO == 1):
        output += 'Go to Position Request\n'

    #gSTA
    output += 'gSTA = ' + str(status.gSTA) + ': '
    if(status.gSTA == 0):
        output += 'Gripper is not activated. See Fault Status if Gripper is activated\n'
    if(status.gSTA == 1):
        output += 'Activation in progress\n'
    if(status.gSTA == 2):
        output += 'Not used\n'
    if(status.gSTA == 3):
        output += 'Gripper is operational\n'

    #gOBJ
    output += 'gOBJ = ' + str(status.gOBJ) + ': '
    if(status.gOBJ == 0):
        output += 'Unknown object detection. Regulating towards requested vacuum/pressure\n'
    if(status.gOBJ == 1):
        output += 'Object detected. Minimum vacuum value reached\n'
    if(status.gOBJ == 2):
        output += 'Object detected. Maximum vacuum value reached\n'
    if(status.gOBJ == 3):
        output += 'No object detected. Object loss, dropped or gripping timeout reached\n'
 
    #gFLT
    output += 'gFLT = ' + str(status.gFLT) + ': '
    if(status.gFLT == 0x00):
        output += 'No Fault\n'
    if(status.gFLT == 0x03):
        output += 'Very porous material detected\n'
    if(status.gFLT == 0x05):
        output += 'Priority Fault: Action delayed, initialization must be completed prior to action\n'
    if(status.gFLT == 0x06):
        output += 'Gripping timeout. rGTO must be re-asserted (rGTO=0 then rGTO=1) or one of the following parameters must be changed (rMOD, rPR, rSP, rFR)\n'
    if(status.gFLT == 0x07):
        output += 'The Activation bit not set. Activation bit must be set prior to action (rACT=1)\n'
    if(status.gFLT == 0x08):
        output += 'Maximum operating temperature exceeded, wait for cool-down\n'
    if(status.gFLT == 0x09):
        output += 'No communication during at least 1 second. This fault will only be returned once if the next valid communication is a "read command" of the FAULT STATUS register\n'

    if(status.gFLT == 0x0A):
        output += 'Under minimum operating voltage\n'
    if(status.gFLT == 0x0B):
        output += 'Automatic release in progress (Vacuum/pressure detected)\n'
    if(status.gFLT == 0x0C):
        output += 'Internal fault; contact support@robotiq.com\n'
    if(status.gFLT == 0x0F):
        output += 'Automatic release completed (Vacuum/pressure not detected)\n' 

    #gPR
    output += 'gPR = ' + str(status.gPR) + ': '
    output += 'Echo of the requested pressure for the vacuum gripper: ' + str(status.gPR) + '\n'

    #gPO
    output += 'gPO = ' + str(status.gPO) + ': '
    output += 'Actual Vacuum/Pressure measured in suction cup: ' + str(100-status.gPO) + '\n'

    return output
    
if __name__ == '__main__':
    RobotiqVacuumGrippersStatusListener()
